StripePaymentSheet/StripePaymentSheet/Source/Internal/Link/Controllers/PayWithLinkWebController.swift

```diff
@@ -0,0 +1,240 @@
+//
+//  PayWithLinkWebController.swift
+//  StripePaymentSheet
+//
+//  Created by Cameron Sabol on 9/3/21.
+//  Copyright © 2021 Stripe, Inc. All rights reserved.
+//
+
+import UIKit
+
+import AuthenticationServices
+@_spi(STP) import StripeCore
+@_spi(STP) import StripePayments
+@_spi(STP) import StripeUICore
+
+@available(iOSApplicationExtension, unavailable)
+@available(macCatalystApplicationExtension, unavailable)
+protocol PayWithLinkWebControllerDelegate: AnyObject {
+
+    func payWithLinkWebControllerDidConfirm(
+        _ payWithLinkWebController: PayWithLinkWebController,
+        intent: Intent,
+        with paymentOption: PaymentOption,
+        completion: @escaping (PaymentSheetResult) -> Void
+    )
+
+    func payWithLinkWebControllerDidCancel(_ payWithLinkWebController: PayWithLinkWebController)
+
+    func payWithLinkWebControllerDidFinish(
+        _ payWithLinkWebController: PayWithLinkWebController,
+        result: PaymentSheetResult
+    )
+
+}
+
+protocol PayWithLinkCoordinating: AnyObject {
+    func confirm(
+        with linkAccount: PaymentSheetLinkAccount,
+        paymentDetails: ConsumerPaymentDetails,
+        completion: @escaping (PaymentSheetResult) -> Void
+    )
+    func confirmWithApplePay()
+    func cancel()
+    func accountUpdated(_ linkAccount: PaymentSheetLinkAccount)
+    func finish(withResult result: PaymentSheetResult)
+    func logout(cancel: Bool)
+}
+
+/// A view controller for paying with Link using ASWebAuthenticationSession.
+///
+/// Instantiate and present this controller when the user chooses to pay with Link.
+/// For internal SDK use only
+
+@available(iOSApplicationExtension, unavailable)
+@available(macCatalystApplicationExtension, unavailable)
+@objc(STP_Internal_PayWithLinkWebController)
+final class PayWithLinkWebController: NSObject, ASWebAuthenticationPresentationContextProviding, STPAuthenticationContext {
+    func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
+        return bestWindowForPresentation
+    }
+    func authenticationPresentingViewController() -> UIViewController {
+        if let presentationVC = presentationVC {
+            return presentationVC
+        }
+        let window = bestWindowForPresentation
+        var presentingViewController: UIViewController = window.rootViewController!
+
+        // Find the most-presented UIViewController
+        while let presented = presentingViewController.presentedViewController {
+            presentingViewController = presented
+        }
+
+        return presentingViewController
+    }
+    var bestWindowForPresentation: UIWindow {
+        if let window = presentationVC?.view.window {
+            return window
+        }
+        if let window = UIApplication.shared.windows.first(where: { $0.isKeyWindow }) {
+            return window
+        }
+        return UIApplication.shared.windows.first!
+    }
+    var presentationVC: UIViewController?
```

Comment:
There's probably a reason we can't do this, but it'd be nice if we could require this in the initializer and make it non-nil